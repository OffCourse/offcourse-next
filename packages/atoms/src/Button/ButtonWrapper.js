import system from "system-components";
import { theme } from "styled-system";

const background = {
  default: "grayScale.3",
  disabled: "disabled",
  primary: "primary",
  positive: "positive",
  warning: "warning",
  negative: "negative"
};

const inverseBackground = {
  default: "primary",
  disabled: "disabled",
  primary: "grayScale.3",
  positive: "grayScale.3",
  warning: "grayScale.3",
  negative: "grayScale.3"
};

const textColor = {
  default: "white",
  disabled: "grayScale.1",
  negative: "white",
  primary: "black",
  warning: "black",
  positive: "black"
};

const inverseText = {
  default: "white",
  disabled: "white",
  negative: "white",
  primary: "white",
  warning: "white",
  positive: "white"
};

const ButtonWrapper = system(
  {
    is: "button",
    display: "flex",
    alignItems: "center",
    height: "2.813rem",
    justifyContent: "center",
    m: 0,
    border: 0,
    fontSize: 1,
    lineHeight: 1,
    minWidth: "5.33333rem",
    maxWidth: "16rem",
    px: 6,
    py: 4,
    borderBottom: 2
  },
  props => ({
    backgroundColor: theme(`colors.${background[props.variant]}`)(props),
    borderColor: theme(`colors.${inverseBackground[props.variant]}`)(props),
    color: theme(`colors.${textColor[props.variant]}`)(props),
    fontFamily: props.theme.fonts.bold,
    "&:focus": { outline: "none" },
    "&:disabled": { cursor: "default" },
    "> a": {
      color: "inherit",
      textDecoration: "inherit"
    },
    "&:hover": {
      borderColor: theme(`colors.${background[props.variant]}`)(props),
      color: theme(`colors.${inverseText[props.variant]}`)(props),
      backgroundColor: theme(`colors.${inverseBackground[props.variant]}`)(
        props
      )
    },
    boxSizing: "border-box",
    userSelect: "none"
  })
);

export default ButtonWrapper;
