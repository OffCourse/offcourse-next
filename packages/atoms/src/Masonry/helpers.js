import {
  append,
  adjust,
  repeat,
  memoizeWith,
  reduce,
  identity,
  addIndex,
  reduceWhile,
  inc
} from "ramda";

const reduceIndexed = addIndex(reduce);

const _prepareGrid = (items, numberOfColumns) => {
  return reduceIndexed(
    (acc, child, index) => adjust(append(child), index % numberOfColumns, acc),
    repeat([], numberOfColumns),
    items
  );
};

const indexGrid = (items, numberOfColumns) => {
  return `${numberOfColumns}-${items.length}`;
};

const prepareGrid = memoizeWith(indexGrid, _prepareGrid);

const _getColumns = (containerWidth, breakpoints) => {
  return reduceWhile(
    (_, breakpoint) => breakpoint < containerWidth,
    inc,
    1,
    breakpoints
  );
};

const getColumns = memoizeWith(identity, _getColumns);

export { getColumns, prepareGrid };
