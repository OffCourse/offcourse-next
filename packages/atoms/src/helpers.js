import titleCase from "voca/title_case";
import lowerCase from "voca/lower_case";
import trimLeft from "voca/trim_left";
import { map, compose, filter, identity } from "ramda";

const titleize = str => titleCase(str, ["'", "-", "â€™"]);

const formatTitle = compose(
  trimLeft,
  titleize,
  lowerCase
);

const compact = filter(identity);

export { compose, map, filter, compact, lowerCase, formatTitle };
