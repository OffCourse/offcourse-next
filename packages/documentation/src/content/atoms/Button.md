Default and Disabled Button

```react|span-2
const clickHandler = () => alert("ALERT");
<Button onClick={clickHandler}>Click Me</Button>
```

```react|span-2
const clickHandler = () => alert("ALERT");
<Button onClick={clickHandler} disabled>
  Cannot Click
</Button>
```

Meaningful variants

```react|span-2
const clickHandler = () => alert("ALERT");
  <Button variant="primary" href="/">
    Click Now!
  </Button>
```

```react|span-2
const clickHandler = () => alert("ALERT");
  <Button variant="positive" onClick={clickHandler}>
    Please Click
  </Button>
```

```react|span-2
const clickHandler = () => alert("ALERT");
  <Button variant="negative" onClick={clickHandler}>
    Don't Click
  </Button>
```

3 different sizes

```react|span-2
const clickHandler = () => alert("ALERT");
  <Button size="small" variant="primary" href="/">
    Click!
  </Button>
```

```react|span-2
const clickHandler = () => alert("ALERT");
  <Button size="medium" variant="positive" onClick={clickHandler}>
    Please Click
  </Button>
```

```react|span-2
const clickHandler = () => alert("ALERT");
  <Button size="large" variant="negative" onClick={clickHandler}>
    Click Click Click Click Click
  </Button>
```
