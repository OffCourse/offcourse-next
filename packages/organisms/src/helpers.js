import titleCase from "voca/title_case";
import lowerCase from "voca/lower_case";
import { compose, filter, identity } from "ramda";

const formatTitle = compose(
  titleCase,
  lowerCase
);
const compact = filter(identity);

export { compact };
