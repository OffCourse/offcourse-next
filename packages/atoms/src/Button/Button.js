import React, { Component } from "react";
import PropTypes from "prop-types";
import ButtonWrapper from "./ButtonWrapper";
import { formatTitle } from "../helpers";

/**
 * The button component for the Offcourse project
 */

class Button extends Component {
  static propTypes = {
    /** determines if the button should be disabled */
    disabled: PropTypes.bool,
    /** determines the status of the button */
    variant: PropTypes.oneOf([
      "default",
      "primary",
      "positive",
      "warning",
      "negative"
    ]),
    /** the text that is displayed on the button */
    children: PropTypes.string,
    /** code that the button should execute when clicked */
    onClick: PropTypes.func,
    /** a url that the button should link to, automatically turns the button into the basic type */
    href: PropTypes.string,
    /** only accept a type prop when type is submit */
    type: PropTypes.oneOf(["submit", "button"]),
    /** the size of the button */
    size: PropTypes.oneOf(["small", "medium", "large"]),
    tabindex: PropTypes.number,
    mt: PropTypes.number
  };

  static defaultProps = {
    variant: "default",
    size: "medium",
    disabled: false,
    type: "button"
  };

  handleClick = e => {
    const { onClick } = this.props;
    e.preventDefault();
    onClick();
  };

  render() {
    const {
      href,
      type,
      children,
      onClick,
      tabindex,
      variant,
      mt,
      disabled,
      size
    } = this.props;

    const buttonType = type || "button";

    const widths = {
      small: "5.33333rem",
      medium: "8rem",
      large: "16rem"
    };

    const text = formatTitle(children);

    return (
      <ButtonWrapper
        mt={mt}
        type={buttonType}
        variant={disabled ? "disabled" : variant}
        disabled={disabled}
        onClick={onClick}
        tabIndex={tabindex || 1}
        minWidth={widths[size]}
        maxWidth={widths[size]}
      >
        {href ? <a href={!disabled ? href : undefined}>{text}</a> : text}
      </ButtonWrapper>
    );
  }
}

export default Button;
