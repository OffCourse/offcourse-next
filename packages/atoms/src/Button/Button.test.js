import '../setupTests';

import React from 'react';
import ReactDOM from 'react-dom';
import { create as render } from 'react-test-renderer'
import { configure, shallow, mount } from 'enzyme';
import sinon from 'sinon';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import Button from './Button';
import ButtonWrapper from './ButtonWrapper';
import { formatTitle } from '../helpers';

describe('Button', () => {
  const props = {
    variant: "primary",
    children: formatTitle("click me!"),
    type: "button",
  };

  it('passes props', () => {
    const tree = shallow(<Button {...props} />);
    
    expect(tree.props()).toMatchObject(props);
  });

  it('creates anchor when href is passed', () => {
    const tree = shallow(<Button href="https://offcourse.io" />);

    expect(tree.find('a')).toHaveLength(1);
    expect(tree.find('a').props().href).toBe('https://offcourse.io');
  });

  it('changes variant prop when disabled', () => {
    const tree = shallow(<Button disabled={true} />);

    expect(tree.props().variant).toBe("disabled");
  });

  it('handles onClick event', () => {
    const onButtonClick = sinon.spy();
    const tree = shallow(<Button onClick={onButtonClick} />);

    tree.simulate('click');

    expect(onButtonClick).toHaveProperty('callCount', 1);
  });

  it('matches the snapshot', () => {
    const tree = shallow(<Button />);

    expect(toJson(tree)).toMatchSnapshot();
  });
});
