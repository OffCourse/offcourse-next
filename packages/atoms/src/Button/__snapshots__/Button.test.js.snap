// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches the snapshot 1`] = `
<Styled(div)
  alignItems="center"
  border={0}
  borderBottom={2}
  display="flex"
  fontSize={1}
  height="2.813rem"
  is="button"
  justifyContent="center"
  lineHeight={1}
  m={0}
  maxWidth="8rem"
  minWidth="8rem"
  mt={0}
  px={6}
  py={4}
  tabIndex={1}
  type="button"
  variant="default"
/>
`;
